ops/testing.py

```diff
@@ -0,0 +1,372 @@
+    # noinspection PyProtectedMember
+    harness._register_charm(the_charm)
+    return harness
+
+
+# noinspection PyProtectedMember
+class TestingHarness:
+    """This class represents a way to build up the model that will drive a test suite.
+
+    The model that is created is from the viewpoint of the charm that you are testing.
+
+    :ivar charm: The instance of Charm that is backed by this testing harness.
+    :type charm: CharmBase
+    """
+
+    def __init__(self, unit_name):
+        """Create a testing harness that can drive a Model"""
+        self.unit_name = unit_name
+        self._backend = _TestingModelBackend(unit_name)
+        self._relation_id_counter = 0
+        self.charm = None
+
+    def _get_backend(self):
+        return self._backend
+
+    @property
+    def model(self):
+        return self.charm.model
+
+    def _register_charm(self, charm):
+        if self.charm is not None:
+            raise RuntimeError(
+                "registering charm {} while {} is already registered".format(
+                    charm, self.charm))
+        self.charm = charm
+
+    def _next_relation_id(self):
+        rel_id = self._relation_id_counter
+        self._relation_id_counter += 1
+        return rel_id
+
+    def add_relation(self, relation_name, remote_app, remote_app_data={}):
+        """Declare that there is a new relation between this app and `remote_app`
+
+        TODO: Once relation_created exists as a Juju hook, it should be triggered by this code.
+
+        :param relation_name: The relation on Charm that is being related to
+        :param remote_app_data: Optional data bag that the remote application is sending
+          If remote_app_data is not empty, this should trigger
+          ``charm.on[relation_name].relation_changed(app)``
+        :return: The relation_id created by this add_relation.
+        :rtype: int
+        """
+        rel_id = self._next_relation_id()
+        self._backend._relation_ids_map.setdefault(relation_name, []).append(rel_id)
+        self._backend._relation_names[rel_id] = relation_name
+        self._backend._relation_list_map[rel_id] = []
+        self._backend._relation_data[rel_id] = {
+            remote_app: remote_app_data,
+            self._backend.unit_name: {},
+            self._backend.app_name: {},
+        }
+        # Reload the relation_ids list
+        self.model.relations._invalidate(relation_name)
+        # TODO: jam 2020-03-05 We should be triggering relation_changed(app) if
+        # remote_app_data isn't empty.
+        return rel_id
+
+    def add_relation_unit(self, relation_id, remote_unit_name, remote_unit_data={}):
+        """Add a new unit to a relation.
+
+        Example::
+
+          rel_id = harness.add_relation('db', 'postgresql')
+          harness.add_relation_unit(rel_id, 'postgresql/0', remote_unit_data={'foo': 'bar'}
+
+        This will trigger a `relation_joined` event and a `relation_changed` event.
```

Comment:
I wonder if we should do this or if we should be a bit more explicit about what is being run. In other words, do we make add_relation_unit emit the events, or do we allow setting up a state, and then asking for a given event to be emitted?

The advantage of doing the former is that the harness will know which specific hooks run in specific events. The disadvantage is that it becomes harder to test specific more scenarios without having to go through a lot of boilerplate first, compared to just allowing people to say "those are the units you have now" and then "fire hook X". Also, how far does it go? Are we calling install, start, etc, when we create the harness? Why not and how does that relate to the choice of sending events automatically here?

These issues are not clear to me yet.. sounds worth discussing it before moving on as this is a behavior we can't change later.

Reply:
An interesting thought presents itself.
If Harness didn't initialize the charm, but you actually called `Harness.enable_events(charm)`.

I think we still need some sort of 'do the framework song and dance', but that could be `charm = harness.initialize(charm_cls)`.
Let me prototype it, and see what it looks like.

Reply:
11:11:59 <Chipaca> niemeyer: any progress on reviewing the testing harness?
11:12:37 <niemeyer> Chipaca: Yeah, I was hoping to cover that today in the standup, but I now realize Jam is out today 
11:12:57 <niemeyer> Chipaca: The most elegant approach seems to be supporting both ends of it
11:13:39 <niemeyer> Chipaca: We might have a couple of options.. one of them is having per method flags, and the other is a more global enable/disable events
11:14:10 <niemeyer> Global is probably better.. not sure about what the default should be
11:14:15 <niemeyer> Maybe emitting?
11:14:33 <niemeyer> Also, it should likely be named emit instead of trigger, as that's what we already use everywhere
11:15:11 <niemeyer> Also shorter, but that's just a bonus of consistency
11:20:44 <Chipaca> :-)
11:20:58 <niemeyer> Not exactly the reply I expected
11:21:14 <Chipaca> >:(
11:21:17 <Chipaca> better?
11:21:49 <Chipaca> niemeyer: sorry, i read it as a status dump, didn't think you were actually asking me there
11:22:49 <Chipaca> niemeyer: I think john was slightly in favour of not emitting by default, but i might have misunderstood. I do think there is value in both approaches, and if we can support both that is the right approach
11:23:09 <niemeyer> Yeah, it seems easy to support both, and indeed it's the best of both worlds
11:23:10 <Chipaca> wrt which to do by default, i don't know which would be more surprising
11:23:36 <niemeyer> The problem with emitting always is that it becomes harder to create specific sequences of events
11:23:37 <Chipaca> the 'set things up, now be alive, now test' seems to flow naturally
11:23:58 <Chipaca> with the other approach being just to ask it to be alive before setting things up
11:24:50 <Chipaca> conversely, 'refrain from emitting, set things up, start emitting, test' seems more awkward
11:24:51 <Chipaca> hm
11:25:12 <Chipaca> niemeyer: so the emit-by-default only makes sense if we're sure that's what people will be using most of the time
11:25:33 <niemeyer> It's a bit unfortunate that we don't have a chance of getting it wrong and then fixing it